pinot-core/src/main/java/com/linkedin/pinot/core/operator/dociditerators/OrDocIdIterator.java

```diff
@@ -15,139 +15,119 @@
+  private final int[] _nextDocIds;
+  private final int _minDocId;
+  private final int _maxDocId;
+
+  private int _numNotExhaustedIterators;
+  private int _currentDocId = -1;
+
+  public OrDocIdIterator(BlockDocIdIterator[] docIdIterators, int minDocId, int maxDocId) {
+    _docIdIterators = docIdIterators;
+    _numNotExhaustedIterators = docIdIterators.length;
+    _nextDocIds = new int[_numNotExhaustedIterators];
+    for (int i = 0; i < _numNotExhaustedIterators; i++) {
+      _nextDocIds[i] = docIdIterators[i].advance(minDocId);
     }
+    _minDocId = minDocId;
+    _maxDocId = maxDocId;
+    removeExhaustedIterators();
   }
 
   @Override
-  public int advance(int targetDocId) {
-    if (currentDocId == Constants.EOF) {
+  public int next() {
+    if (_currentDocId == Constants.EOF) {
       return Constants.EOF;
     }
-    if (targetDocId < minDocId) {
-      targetDocId = minDocId;
-    } else if (targetDocId > maxDocId) {
-      currentDocId = Constants.EOF;
-      return currentDocId;
-    }
-    long start = System.nanoTime();
 
-    // Remove iterators that are before the target document id from the queue
-    Iterator<IntPair> iterator = queue.iterator();
-    while (iterator.hasNext()) {
-      IntPair pair = iterator.next();
-      if (pair.getLeft() < targetDocId) {
-        iterator.remove();
-        iteratorIsInQueue[pair.getRight()] = false;
+    int nextDocId = Integer.MAX_VALUE;
+    boolean hasExhaustedIterator = false;
+    for (int i = 0; i < _numNotExhaustedIterators; i++) {
+      int docId = _nextDocIds[i];
+      if (docId == _currentDocId) {
+        docId = _docIdIterators[i].next();
+        _nextDocIds[i] = docId;
       }
-    }
-
-    // Advance all iterators that are not in the queue to the target document id
-    for (int i = 0; i < docIdIterators.length; i++) {
-      if (!iteratorIsInQueue[i]) {
-        int nextDocId = docIdIterators[i].advance(targetDocId);
-        if (nextDocId != Constants.EOF) {
-          pointers[i].setLeft(nextDocId);
-          queue.add(pointers[i]);
-        }
-        iteratorIsInQueue[i] = true;
+      if (docId != Constants.EOF) {
+        nextDocId = Math.min(nextDocId, docId);
+      } else {
+        hasExhaustedIterator = true;
       }
     }
-
-    // Return the first element
-    if (queue.size() > 0) {
-      currentDocId = queue.peek().getLeft();
+    if (nextDocId > _maxDocId) {
+      _currentDocId = Constants.EOF;
     } else {
-      currentDocId = Constants.EOF;
+      _currentDocId = nextDocId;
+      if (hasExhaustedIterator) {
+        removeExhaustedIterators();
+      }
     }
-
-    long end = System.nanoTime();
-    timeMeasure.addAndGet(end - start);
-    return currentDocId;
+    return _currentDocId;
   }
 
   @Override
-  public int next() {
-    long start = System.currentTimeMillis();
-
-    if (currentDocId == Constants.EOF) {
-      return currentDocId;
+  public int advance(int targetDocId) {
```

Comment:
what is the contract of advance, find matching element that is greater than or equal to targetDocId or greater than targetDocId

Reply:
Yes, will add it into the interface